### PR TITLE
[JUJU-4099] Add watcher to upgrade that notifies when upgrade can start

### PR DESCRIPTION
**Pull request by jack-w-shaw**
Add watcher to upgrade that notifies when upgrade can start

Instead of, as we are with mongo, watching for any changes and then
requiring the consumer to verify it's an actionable change, create a new
watcher extending NamespaceWatcher which does this verification for you

This combines Watch and AllProvisionedControllersReady from
state/upgrade

Test this by mocking the underlying watcher. Ideally, we would have some
tests watching a real database, but we are not able to do this at the
moment.

## Checklist

- [x] Code style: imports ordered, good names, simple structure, etc
- [x] Comments saying why design decisions were made
- [x] Go unit tests, with comments saying what you're testing
- ~[ ] [Integration tests](https://github.com/juju/juju/tree/main/tests), with comments saying what you're testing~
- [x] [doc.go](https://discourse.charmhub.io/t/readme-in-packages/451) added or updated in changed packages

## QA steps

```sh
go test github.com/juju/juju/domain/upgrade/...
```

### CODE REVIEW

**Review by manadart — COMMENTED** [review]
Looks good; just need to move the base watcher instantiation.

File: domain/upgrade/service/watcher.go [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+		upgradeUUID: upgradeUUID,
+		out:         make(chan struct{}),
+		in:          namespaceWatcher.Changes(),
+	}
+
+	if err := catacomb.Invoke(catacomb.Plan{
+		Site: &w.catacomb,
+		Work: w.loop,
+	}); err != nil {
+		return nil, errors.Trace(err)
```
[the remark this answers is not in the document]
Comment (jack-w-shaw) [reply domain/upgrade/service/watcher.go:47]: Good spot! Fixed

File: domain/upgrade/service/watcher.go [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+	defer close(w.out)
+
+	namespaceWatcher, err := w.wf.NewNamespaceWatcher("upgrade_info_controller_node", changestream.Create|changestream.Update, "")
+	if err != nil {
+		return errors.Trace(err)
+	}
+	if err := w.catacomb.Add(namespaceWatcher); err != nil {
+		return errors.Trace(err)
+	}
+	w.in = namespaceWatcher.Changes()
```
Comment (manadart) [inline domain/upgrade/service/watcher.go:71]: Unlike `w.out`, this is only used in the loop. You don't actually need a member for it.
Just assign the local `in` to either `namespaceWatcher.Changes()` or `nil` based on tick/tock.

**Review by manadart — APPROVED** [review]
Good to go with one more small change.

File: domain/upgrade/service/watcher.go (hunk 2 above)
Comment (jack-w-shaw) [reply domain/upgrade/service/watcher.go]: Ah yes. This redundancy is an artifact from earlier code I believe

**Pull request merged**